clang/lib/Sema/SemaDecl.cpp

```diff
@@ -803,9 +803,13 @@ void Sema::DiagnoseUnknownTypeName(IdentifierInfo *&II,
   // (struct, union, enum) from Parser::ParseImplicitInt here, instead?
 
   if (!SS || (!SS->isSet() && !SS->isInvalid()))
-    Diag(IILoc, IsTemplateName ? diag::err_no_template
-                               : diag::err_unknown_typename)
-        << II;
+    if (getLangOpts().SYCLIsDevice && !getLangOpts().GPURelocatableDeviceCode &&
+        II->getName() == "SYCL_EXTERNAL")
+      Diag(IILoc, diag::err_sycl_external_no_rdc);
+    else
+      Diag(IILoc,
+           IsTemplateName ? diag::err_no_template : diag::err_unknown_typename)
```

Comment:
What if there is `SYCL_EXTERNAL` AND unknown type name?

Reply:
Thanks for the review. I think in general that yes it should be up to the user to check before using it. However if someone isn't familiar with the SYCL spec they might not know why they are getting a missing type error, so the specific error has some value, but it definitely may not be worth it given the gross implementation.

Sorry, can you explain the case you're worried about? My understanding of my change is when it's about to throw the usual unknown type error, if the unknown type name == `SYCL_EXTERNAL`, throw a different error. However it's definitely possible the change has an unintended side effort.

Reply:
> Sorry, can you explain the case you're worried about?

I was wondering about the cases like
```
SYCL_EXTERNAL bla foo(); // bla is an intended error.
bla SYCL_EXTERNAL foo();
```
How this patch changes existing behavior for them?

Reply:
The diagnostic is clearer for the first case and unchanged for the second.  I am okay with this.

Reply:
I agree the diagnostic is better but I do not know if this change will  be accepted by community clang if/when we do upstream since it is a bit hacky. @erichkeane can you please weigh in here? Is this acceptable?

Reply:
yeah im happy to drop this part if its too gross, thanks

Reply:
While the diagnostic is better, I don't think this is the sort of thing that community would approve of.  Recognizing it by name exactly like that is pretty objectionable.

Reply:
alright ill remove this, thanks for the discussion